core/src/crds_gossip_pull.rs

```diff
@@ -391,25 +409,27 @@ impl CrdsGossipPull {
     // build a set of filters of the current crds table
     // num_filters - used to increase the likelyhood of a value in crds being added to some filter
     pub fn build_crds_filters(&self, crds: &Crds, bloom_size: usize) -> Vec<CrdsFilter> {
+        const PAR_MIN_LENGTH: usize = 1024;
         let num = cmp::max(
             CRDS_GOSSIP_DEFAULT_BLOOM_ITEMS,
             crds.table.values().count() + self.purged_values.len(),
         );
-        let mut filters = CrdsFilterSet::new(num, bloom_size);
-        let mut add_value_hash = |value_hash| {
-            if let Some(filter) = filters.get(value_hash) {
-                debug_assert!(filter.test_mask(value_hash));
-                filter.filter.add(value_hash);
-            }
-        };
-        for v in crds.table.values() {
-            add_value_hash(&v.value_hash);
-        }
-        for (value_hash, _insert_timestamp) in &self.purged_values {
-            add_value_hash(&value_hash);
-        }
-
-        filters.0
+        let filters = CrdsFilterSet::new(num, bloom_size);
+        rayon::join(
+            || {
+                crds.table
+                    .par_values()
+                    .with_min_len(PAR_MIN_LENGTH)
+                    .for_each(|v| filters.add(v.value_hash))
```

Comment:
Does aggregating the values for each filter outside of the lock before calling `add` to bulk insert those values improve the bench at all?

Reply:
Initially I tried something similar to what I think you are suggesting, and it was slightly slower. I have some ideas about a lock-free implementation of the bloom filter construction, that I am going to test as well and compare.

Reply:
Yea, I was thinking of doing like a batch calculate a set of positions to set outside the lock.. then go into the lock and bulk set the positions. It might be more beneficial on a larger machine with more cores which would show more lock contention.

Reply:
sent out https://github.com/solana-labs/solana/pull/12422 which adds an atomic variant of the bloom filter, and will use it here once that is merged. so wouldn't need locking and mutex here anymore.